### ISSUE TITLE
Implement FacturX/XRechnung/ZUGFeRD for upcoming obligation to issue and receive electronic invoices

### ISSUE THREAD

**Issue by bastolino** [Dolibarr/dolibarr#30078]
### Feature Request

As already asked by https://github.com/Dolibarr/dolibarr/issues/8836
FacturX/XRechnung is a european standard (EU-Richtlinie 2014/55/EU / EN 16931) and being enforced between companies from 2025 on, it is growing a basic feature that'd rule out the usage of Dolibarr in Germany from 2027 on. It is expectable that it is going to be mandatory for other european countries aswell.

It should be possible to import invoices from FacturX aswell as import format for supplier invoices.

### Use case

all companies in germany need this from 2027 on.

### Suggested implementation

some hints have already been made regarding libraries.

### Suggested steps

xml embedded in pdf for export,
xml aswell as pdf w/ embedded xml for import

**Comment by Adventurousdonut** [comment]
To add to that, here are some ressources for xml invoices from a subsidiary of the Austrian Chamber of Commerce: https://github.com/austriapro/ebinterface-standards

**Comment by rycks** [comment]
@bastolino  and all others, please note that there is a module for that : https://www.dolistore.com/fr/crm-gestion-relation-client/1511-FacturX---Facture---lectronique--Chorus----.html

i'm a the official developer of that module and i explain that this module will become free (gratuit / gratis / 0€) when that format will be mandatory (please read that for example: https://www.dolibarr.fr/forum/t/est-ce-que-dolibarr-pourra-lire-les-facture-x-en-2024/43169/9 in french but please use online translators if you can't read french).

All users who pay for that module make a direct sponsoring action because it's a hard job to make that sort of module (please note i do 6 "major" release and near that 100 intermediate releases).

So don't waste your time and let me know if you want to contribute to that module (source code is GNU/GPL).

**Comment by Basti-Fantasti** [comment]
@rycks Thanks for your response. This is great news that you will make this module freely available

As it will be mandatory for all companies in the EU it would be good, if this could be added as an inbuilt / default module in dolibarr.

**Comment by rycks** [comment]
@Basti-Fantasti maybe you have an information i don't but for me FacturX/XRechnung is only for Germany and France, not the others countries in europe: they use PeppolXML files / UBL 

And even if that file format was for all european people i think it's a bad idea to put it into main dolibarr code because of all the rest of the world don't care about that specific file format.

For example do we have to put into dolibarr main code specific file format for each country / region ? We are working on a separate repository with external modules with community support which seems to be more flexible (and modules could be more "up to date" / "updates released faster" than the core).

**Comment by Basti-Fantasti** [comment]
@rycks oh I see. I thought that all European countries would need to have the FacturX/XRechnung implemented.

You are also right that it would not make much sense to add it to the default dolibarr installation from that point of view.
But how would the implementation then work? Would everyone need to download the module from dolistore or would there be another way of distributing this module?

**Comment by bastolino** [comment]
Correct me but isn't this all very similiar with some changes to suit each countries flavour? FacturX/XRechnung is basically planned to be used within Peppol (Pan-European Public Procurement Online) Network for delivery.
As far as I understand it, Peppol is the data exchange platform and FacturX/XRechnung is the format.

There is a (german) technology blog by SAP which brings all that into context:
https://community.sap.com/t5/technology-blogs-by-sap/xrechnung-zugferd-und-peppol-alles-was-sie-%C3%BCber-die-elektronische-rechnung/ba-p/13427891

**Comment by gerardkieffer** [comment]
I often read in Dolibarr forums that electronic invoicing will become mandatory in 2026/27. This is the date for France! Other EU countries have different deadlines. In Luxembourg, where I am, many types of companies (not all yet) are already obliged to use the Peppol network to send and receive invoices since March 2023.
Therefore, please keep in mind that there are Dolibarr users for whom the e-invoicing obligation, and the obligation to use Peppol are already in place. 

@bastolino, you're correct: Peppol is the network, and Factur-X is an e-invoice format. The used formats vary by country. For example, France accepts Factur-X and UBL (among others). In Luxembourg, we use XML, UBL, or something called CII, I think. The Peppol access points convert the sent and received information (invoices and other documnets) from one format to another. Peppol access points communicate using "messages" in a common format.

**Comment by bastolino** [comment]
Yes. In Germany, being able to receive electronic invoices is mandatory from January 1st 2025 on for each and every company. The need to be able to send electronic invoices is mandatory depending on company size.

The big point is that - for Germany - from 2025 on the XML part of the invoice is the leading one if there are differences between the (human readable) PDF and the XML part. While there is already a (buyable) Module for exporting FacturX, I don't see the possibility of parsing the XML part of it.

So is it possible to create an import tool to parse the FacturX and simply use the XML part of it to create a supplier invoice?

**Comment by rycks** [comment]
@bastolino import invoices is the job of ScanInvoices (see on dolistore) :-) and in case of factur-x import is done full automatic :-)

yes dolibarr is ready (and some of us are doing with that) !

**Comment by bastolino** [comment]
@rycks yes, but to use that you need an abo with your company and each and every invoice is sent to your webservers. importing pdfs with xmls shouldn't make OCR necessary, especially not sending them to any cloud service hosted by a different company.

**Comment by spoonerarthur** [comment]
01.01.2025 is getting closer and closer.
Every company must guarantee the receipt of an XML invoice and also the automatic processing in the accounting department.
If I have understood correctly, a special e-mail address must be set up for this, which Dolibarr must then retrieve and process.
At the moment I don't know how it works in Dolibarr.

**Comment by spoonerarthur** [comment]
I've now done some more research on the internet.
I think companies will have to fulfil both requirements, receiving and sending e-invoices

**Comment by bastolino** [comment]
@spoonerarthur 
as far as I understand: it's not yet implemented at all.
The rules are: from Jan 1st 2025 on every company in Germany is required to be able to receive e-invoices. Depending on their size they will need to send e-invoices by 2027.
For sending e-invoices, the FacturX-module is capable, with the latest updates a log of bugs and translational issues are solved.
Regarding the receival process: I don't see any steps forward. The ocr stuff from Eric is OCR only and therefor not sufficient for the german legislature on XRechnung regarding xml first.
(that means: if there are differences between pdf viewable and its corresponding xml, xml has priority for the fiscal authorities)

**Comment by hregis** [comment]
@spoonerarthur @Basti-Fantasti
and Germany will not be conciliatory if a company is not yet in order by January 1, 2025?
You're all going to end up in prison on January 1, do you think?
I pity you! ;-)
In my opinion, you have to stay calm, @rycks will sort out the problem, I trust him!

**Comment by hregis** [comment]
You have to tell your German government and all the governments of Europe that it is all well and good to be part of Europe, but if each country makes its own soup, Europe makes no sense!

**Comment by hregis** [comment]
To put it simply, for our society which is becoming stupid... 
it's like a football team where each player has their own rules of the game... 
it's a mess! 
And Europe is rubbish!

**Comment by rycks** [comment]
Hello @bastolino 

> Regarding the receival process: I don't see any steps forward. The ocr stuff from Eric is OCR only and therefor not sufficient for the german legislature on XRechnung regarding xml first.
> (that means: if there are differences between pdf viewable and its corresponding xml, xml has priority for the fiscal authorities)

please have a look at https://www.dolistore.com/7137-thickbox_default/OCR---ScanInvoices.jpg

ScanInvoices process is:
* try to find xml data, if present use it, if conform to factur-x import that
* fallback to data extraction if no xml
* fallback to OCR if nothing works before

So you can import factur-x invoices sinces 2 years :-) if you want to check it i could make you an invoice just let me know the amount you want :-)

**Comment by rycks** [comment]
@spoonerarthur 

> If I have understood correctly, a special e-mail address must be set up for this, which Dolibarr must then retrieve and process.
> At the moment I don't know how it works in Dolibarr.

here is the answer (in french) but let me know if somebody want to translate it in other languages, i will copy all of that on dolibarr wiki to make translations easy !

https://doc.cap-rel.fr/projet_scaninvoices/import_mail

**Comment by bastolino** [comment]
Hello @rycks 
I have tried out your service with some dolibarr test instance and some real invoices I received containing xml. It did not work and the amounts were just clearly wrong. Also textfields were cut off and some other minor errors that happen when using OCR.
Also, I don't want to upload all of my invoices to a third party and pay per invoice. Then I could go ahead and use DATEV aswell. But I want it on my own premise with as much open source as possible. I totally agree that it must not be for free but not that way and not to have such errors when trying just 3 invoices.

**Comment by rycks** [comment]
@bastolino so that's bugs ... did you open a ticket to send us that invoices to make analyze and correct bugs ? please do it

ScanInvoices is AGPL 

Server code name is "docwizon" (it convert documents to json data) and docwizon could be self hosted like some of our partners does, i'm a fan of self hosted systems (and i make only free software since 1998 ...) so no problem with that.

**Comment by sonikf** [comment]
Hi everyone
@eldy @rycks @hregis and Dolibarr team

This is primarily related to #16898

I think we are in a miss the forest for the trees situation..


In simple words **Peppol** is the network you have to connect **regardless of the format** all through an access point to **send and receive invoices** and possibly other documents **automatically**(this is the magic word)

As @RetroYogi(only one got it right so far) correctly stated Electronic invoicing is already implemented to various degrees all over the world with Peppol being the dominant standard so far.
Companies from countries outside Europe such as Australia, Canada, China, India, Japan, USA are utilizing Peppol for cross-border trade exchanges.
You can see the full list [here](https://peppol.org/members/peppol-certified-service-providers/)

All major commercial and open source ERP's have already implemented it as a **core** feature because simply Electronic invoicing is the new invoicing.

Please see [Odoo implementation ](https://www.odoo.com/documentation/17.0/applications/finance/accounting/customer_invoices/electronic_invoicing.html)  offered since October 2021

 
@rycks 
To start thank you for your work and contributions.
In light of the above please be kind and answer some questions

- Will you also give for free https://www.dolistore.com/en/modules/1585-Peppol-XML---facture-electronique.html ?

- If yes my understanding is that in contrast with the FacturX module you only create the xml in Peppol module, what about sending/transmission? (In Dolistore you provide changelog only in French but in English, German  info is scarce and points to only xml generation please fix this, you maybe losing customers!)
Bear in mind that vendor lock-ins have already been imposed for access points providers.
For example in Greece where Electronic invoicing is mandatory for B2G transactions only a Greek ministry licenced authorized access point provider can be used and a third country provider can either register as a Greek company and receive a licence or connect through a Greek access point provider(i hope this change for B2B or a platform like chorus will be created)
So software bridges will be needed and this is also a new stream of revenue for module developers but also extra costs for customers..

- Your module https://www.dolistore.com/7137-thickbox_default/OCR---ScanInvoices.jpg is only necessary for FacturX because in your module you only use chorus if i got that right by reading your changelogs.
What about Peppol formats where you can request data the same way you transmit and automatically import vendor invoices that way?

-  You wrote "We are working on a separate repository with external modules with community support which seems to be more flexible (and modules could be more "up to date" / "updates released faster" than the core)."
A NEXTCLOUD like implementation with module auto update would be perfect!
Where is this repo at (i can't find it ) and how can i or anyone else willing can help?


-  Can you please provide a timeline/roadmap?

To address @hregis comments: Also thanks for your work and i agree with you that Europe is a mess but in this case France and Germany changed the recipe reinventing the wheel with FacturX/XRechnung/ZUGFeRD
No one will go to prison and i also trust @rycks will figure this out but this is business software. Governments already offer incentives for early adoption and the fact is that unfortunately Dolibarr can't reliably provide right now!

**Comment by eldy** [comment]
There is a wiki page to describe the situation for each country here

https://wiki.dolibarr.org/index.php?title=Electronic_Billing

If you have information about format, dates of expectation for ebilling, thanks to fill this page...

**Comment by sonikf** [comment]
Hi @eldy 
Thanks for creating the wiki page!
I have amended the already created entries for cohesion and structure.
Also I added e-reporting which is also coming your way(France) in same timeline as e-invoicing.
This is the same as  MyData module used by the Greek community to transmit xml invoice data via an API.
Still work to be done..
After speaking to an invoicing expert, extensively reading French legislation and to some extent German and visiting accredited French PDP's websites i verified that both countries officially support Peppol if not for other reasons for cross border exchanges.
So is there a plan for Dolibarr to support Peppol?

**Comment by eldy** [comment]
> Hi @eldy 
> Thanks for creating the wiki page!
> I have amended the already created entries for cohesion and structure.
> Also I added e-reporting which is also coming your way(France) in same timeline as e-invoicing.
> This is the same as  MyData module used by the Greek community to transmit xml invoice data via an API.
> Still work to be done..
> After speaking to an invoicing expert, extensively reading French legislation and to some extent German and visiting accredited French PDP's websites i verified that both countries officially support Peppol if not for other reasons for cross border exchanges.
> So is there a plan for Dolibarr to support Peppol?
> 
> 

Thanks. What you mean with the text "e-reporting" is related to the transmission or declaration of the invoice (sending or transfer), right ?

I agree with you that the faster we have support natively or from a centralized community repository (not yet available) instead of support by an non official module, the better it is.
But roadmap is above all handled by submitted contributions of developers. 
So I can't say myse if there is plan or not. Just that there is hope or intent by some people.
In all cases, I will discuss about Peppol or other support during next devcamp in november to raise awareness to progress on this topics ...

**Comment by sonikf** [comment]
> Thanks. What you mean with the text "e-reporting" is related to the transmission or declaration of the invoice (sending or transfer), right ?


No this is mainly for tax evasion and vat fraud checks from central authorities, please read section "Digital reporting requirements" from https://ec.europa.eu/digital-building-blocks/sites/display/DIGITAL/eInvoicing+in+France
See our rudimentary but working implementation https://github.com/evansgl/dolibarr-mydata/tree/dolibarr-19/mydata/mydatalist

API easy translation with firefox https://www.aade.gr/sites/default/files/2024-07/myDATA%20API%20Documentation%20v1.0.9_official_erp.pdf

> I agree with you that the faster we have support natively or from a centralized community repository (not yet available) instead of support by an non official module, the better it is.
> But roadmap is above all handled by submitted contributions of developers.
> So I can't say myse if there is plan or not. Just that there is hope or intent by some people.
> In all cases, I will discuss about Peppol or other support during next devcamp in november to raise awareness to progress on this topics ...


Thank you 
I'll open another issue about roadmap because i see evolution of parts but no communication is done before, so willing Dolibarrians can't offer insight or help. I think by being open and communicating issues and plans more people will help. I saw this issue by chance searching for sth else by recently updated but not everyone does that...
In this spirit a request if possible, to open polls in forums

**Comment by ultrasites** [comment]
@eldy Maybe we could implement a PR with this dependency and some developers? https://github.com/horstoeko/zugferd

**Comment by rycks** [comment]
@ultrasites of course, that repo is already included into the facturx module and horstoeko do a incredible work but please read what we explain before, facturx is only for german and french people, then we did not include 20Mb of libs into dolibarr core...

@sonikf yes we are working on the community-official repository, we speaked about it during last week (devcamp in nancy) with @eldy and eldy asked me to manage (or be a part of managers of course) for that repository, nice i will do my best and it will take time (not paid)

@sonikf roadmap without $$$ is a little bit hard to plan no ? writing free (gpl) software is my job then i have to find solution to pay my time, dolistore is the way we use up to now but what for the future ?

Next step (yes a roadmap !!!!!) for me seems to be:

- a) set up dolibarr-community repository
- b) make a proposition on the wiki linked to that repository for the best way to put data (one folder per module, then one branch per dolibarr main version, for example FacturX branch 18.0 ... FacturX branch 20.0 but facturx or xrechnung please do not make a german/french fight then facturx-xrechnung folder... ?)
- c) put code into it and all developper documentation (how to build that module with external depends and dolibarr "composer" point of view)
- d) make release process
- e) add dolibarr core auto-connexion to it
- f) tests & validate (with dolibarr experiment tag)
- g) put into dolibarr main without experiment tag

Then ... during that time you can buy the module (and sponsoring the job with that, please note that 20% of the amount is to dolibarr association) and contribute to "my" git repository (all of "my" customers has a access to that repository and can contribute, some of them do it like you can see on the changelog of facturx module for example).

**Comment by ultrasites** [comment]
> @ultrasites of course, that repo is already included into the facturx module and horstoeko do a incredible work but please read what we explain before, facturx is only for german and french people, then we did not include 20Mb of libs into dolibarr core...
> 
> @sonikf yes we are working on the community-official repository, we speaked about it during last week (devcamp in nancy) with @eldy and eldy asked me to manage (or be a part of managers of course) for that repository, nice i will do my best and it will take time (not paid)
> 
> @sonikf roadmap without $$$ is a little bit hard to plan no ? writing free (gpl) software is my job then i have to find solution to pay my time, dolistore is the way we use up to now but what for the future ?
> 
> Next step (yes a roadmap !!!!!) for me seems to be:
> 
>     * a) set up dolibarr-community repository
> 
>     * b) make a proposition on the wiki linked to that repository for the best way to put data (one folder per module, then one branch per dolibarr main version, for example FacturX branch 18.0 ... FacturX branch 20.0 but facturx or xrechnung please do not make a german/french fight then facturx-xrechnung folder... ?)
> 
>     * c) put code into it and all developper documentation (how to build that module with external depends and dolibarr "composer" point of view)
> 
>     * d) make release process
> 
>     * e) add dolibarr core auto-connexion to it
> 
>     * f) tests & validate (with dolibarr experiment tag)
> 
>     * g) put into dolibarr main without experiment tag
> 
> 
> Then ... during that time you can buy the module (and sponsoring the job with that, please note that 20% of the amount is to dolibarr association) and contribute to "my" git repository (all of "my" customers has a access to that repository and can contribute, some of them do it like you can see on the changelog of facturx module for example).

Nice news!

When you will start?

**Comment by carstengrimm** [comment]
There are three weeks until the End of the Year. I think time is running out to make a lot of propositions and decide whether or not it will be added as an additional module or core feature.

**Comment by rycks** [comment]
Note: dolibarr community repository is available : https://github.com/Dolibarr/dolibarr-community-modules

**Comment by John-H-Smith** [comment]
@rycks Nice, thanks! How do I have to install it?

**Comment by Basti-Fantasti** [comment]
@rycks Is there some documenation available on how to setup / use the dolibarr community modules?

When looking at the project it only seems to be a more or less a stub.
The only module source included seems to be the "helloasso"
But also there is no documentation on how to add it to the dolibarr installation.
It only refers to a module zip file for installation.

The other modules only reference their own git repositories
Thanks in advance

**Comment by rycks** [comment]
@Basti-Fantasti the roadmap is detailed on my previous message on that thread (https://github.com/Dolibarr/dolibarr/issues/30078#issuecomment-2490289128)!

[x] a) set up dolibarr-community repository

now there is a tons of questions about step b)

[ ] b) make a proposition on the wiki linked to that repository for the best way to put data (one folder per module, then one branch per dolibarr main version, for example FacturX branch 18.0 ... FacturX branch 20.0 but facturx or xrechnung please do not make a german/french fight then facturx-xrechnung folder... ?)

because if we put all source code into one repository we could not make "split" for managers / reviewers and the question was : it could be more usable if each community module has his own repository and the main repository store "links to" the others

then official (main) dolibarr developpers have to manage the "index / directory" of official modules, and each module could have his own organisation

for example odoo community made the choice "one module = one repository"

what do you (not only "you" @Basti-Fantasti, but all of community members) think about that ?

**Comment by bastolino** [comment]
Appreciate your kindness, @rycks!
Still, the receival process is open and in my eyes, it should be only the xml to be interpreted. (And we need to be able to extract the xml from pdf)
German government uses this project for validation at xrechnung.bund.de: https://github.com/itplr-kosit/validator

**Comment by Basti-Fantasti** [comment]
Hi @rycks and thanks for your quick response.

Sorry I forgot about the roadmap shown above. 
You are right and the upcoming steps need to be discussed on how to use that community module approach in the future.

**Comment by ultrasites** [comment]
Maybe is it possible to download the module from https://inligit.fr/cap-rel/dolibarr/plugin-facturx as ZIP and include it into your own dolibarr instance?

**Comment by shasta123** [comment]
First of all, thank you @rycks for making the module available to the public.

I'm trying to test the module on Debian 12.9 (php 8.2), with doli 20.0.4  but I still got some problems.
Here is what I did so far:

* Get php-scoper from https://github.com/humbug/php-scoper/releases/tag/0.18.14
* Move php-scoper.phar to /usr/local/bin

* Get phpstan from https://github.com/phpstan/phpstan/releases (2.1.1)
* Move phpstan.phar to /usr/local/bin

* Install composer (2.5.5-1+deb12u2)

As I use php 8.2 I changed "php7.4" in Makefile.local to "php". I also had to change
"php7.4 /usr/local/bin/composer.phar" to "composer"

* composer i
* make zip

While php-scoper.phar seems to work, I got an error running phpstan:
(cd "/tmp/facturx-scoper" && phpstan.phar analyse --error-format gitlab --no-progress -c .phpstan.local.neon --memory-limit 6G)
PHP Fatal error:  Uncaught TypeError: PHPStan\File\FileExcluder::isAbsolutePath(): Argument #1 ($path) must be of type string, _PHPStan_2f712479f\Nette\DI\DynamicParameter given, called in phar:///usr/local/bin/phpstan.phar/src/DependencyInjection/ValidateExcludePathsExtension.php on line 41 and defined in phar:///usr/local/bin/phpstan.phar/src/File/FileExcluder.php:102

Commenting out the phpstan line in the local Makefile a zip file is created:

(cd "/tmp/facturx-scoper" && php build/buildzip.php)
extract data from core/modules/modFacturx.class.php
module name = facturx, version = 1.6.51
module archive is ready : /tmp/module_facturx-1.6.51.zip ...

I could install the zip file to dolibarr and enable the module.

When creating an invoice, however, I get an PHP error:

[Mon Jan 20 10:59:31.653307 2025] [proxy_fcgi:error] [pid 3639489:tid 3639514] [remote 78.94.143.90:58364] AH01071: Got error 'PHP message: PHP Fatal error:  Uncaught Error: Call to a member function setExchangedDocumentContext() on null in /export/vhost/dolibarr.mb-net.net/htdocs/custom/facturx/vendor/horstoeko/zugferd/src/ZugferdObjectHelper.php:476\nStack trace:\n#0 /export/vhost/dolibarr.mb-net.net/htdocs/custom/facturx/vendor/horstoeko/zugferd/src/ZugferdDocument.php(95): custom\\facturx\\horstoeko\\zugferd\\ZugferdObjectHelper->getCrossIndustryInvoice()\n#1 /export/vhost/dolibarr.mb-net.net/htdocs/custom/facturx/vendor/horstoeko/zugferd/src/ZugferdDocumentBuilder.php(91): custom\\facturx\\horstoeko\\zugferd\\ZugferdDocument->createInvoiceObject()\n#2 /export/vhost/dolibarr.mb-net.net/htdocs/custom/facturx/vendor/horstoeko/zugferd/src/ZugferdDocumentBuilder.php(82): custom\\facturx\\horstoeko\\zugferd\\ZugferdDocumentBuilder->initNewDocument()\n#3 /export/vhost/dolibarr.mb-net.net/htdocs/custom/facturx/class/facturx.class.php(293): custom\\facturx\\horstoeko\\zugferd\\ZugferdDocumentBuilder::createNew()\n#4 /export/vhost/dolibarr.mb-net.n...', referer: https://dolibarr.mb-net.net/compta/facture/card.php?id=2&save_lastsearch_values=1

Will the module work with php 8.2 or do I have to use 7.4?

As for your question a couple of days ago, I'd prefer the "one module = one repository" approach, with branches for the dolibarr main version.

Cheers,
Michael

**Comment by rycks** [comment]
@shasta123 (and others) for such bug repport please use https://inligit.fr/cap-rel/dolibarr/plugin-facturx

i've just started dev.md documentation some days ago for people like you then you could make bug reports, PR and other hacks on that repo : this is a public AND **community** repository :)

**Comment by Basti-Fantasti** [comment]
@shasta123 with php 8.2 I'm not sure. I have tested it on my ubuntu with php 8.1 and it worked after modifying the Makefile.dist to use this version

**Comment by shasta123** [comment]
Hi @Basti-Fantasti,

Basically I used the wrong version of php-scoper. The current dev.md in Rycks Repository does now contain a tad more detailed description of the auxiliary tools used to build the module.

**Comment by ultrasites** [comment]
Hello,

my solution is via dockerfile.

1. Clone repository
2. Touch dockerfile in new dir
3. Make subdir "app"
4. Copy reposiroty into the workdir "app" 
5. Build docker file
6. Copy the plugin *.zip via docker exec from /tmp to the workdir /app

Here is my dockerfile:

```
FROM debian:12.9

# Install necessary dependencies
RUN apt-get update && apt-get install -y \
    curl \
    unzip \
    php-cli \
    php-gd \
    php-xml \
    php-zip \
    composer \
    make \
    && rm -rf /var/lib/apt/lists/*

# Install php-scoper
RUN curl -L -o /usr/local/bin/php-scoper \
    https://github.com/humbug/php-scoper/releases/download/0.18.11/php-scoper.phar \
    && chmod +x /usr/local/bin/php-scoper

# Install phpstan
RUN curl -L -o /usr/local/bin/phpstan \
    https://github.com/phpstan/phpstan/releases/download/1.12.15/phpstan.phar \
    && chmod +x /usr/local/bin/phpstan

# Verify installation
RUN php-scoper --version && phpstan --version

# Define a volume for persistent storage
VOLUME ["/app"]

# Set working directory
WORKDIR /app


# Install dependencies with composer
RUN composer install --no-interaction --prefer-dist

# Run make zip
RUN make zip
```

Enjoy! ;)
@rycks maybe you can add it in the readme.

**Comment by carstengrimm** [comment]
> Hello,
> 
> my solution is via dockerfile.
> 
> 1. Clone repository
> 2. Touch dockerfile in new dir
> 3. Make subdir "app"
> 4. Copy reposiroty into the workdir "app"
> 5. Build docker file
> 6. Copy the plugin *.zip via docker exec from /tmp to the workdir /app
> 

can you elaborate a bit more? 

- you provide a dolibarr installation into the package??
- or are you adding the "app" folder and dockerfile to your dolibarr root then run dockerfile?

**Comment by ultrasites** [comment]
> > Hello,
> > my solution is via dockerfile.
> > 
> > 1. Clone repository
> > 2. Touch dockerfile in new dir
> > 3. Make subdir "app"
> > 4. Copy reposiroty into the workdir "app"
> > 5. Build docker file
> > 6. Copy the plugin *.zip via docker exec from /tmp to the workdir /app
> 
> can you elaborate a bit more?
> 
>     * you provide a dolibarr installation into the package??
> 
>     * or are you adding the "app" folder and dockerfile to your dolibarr root then run dockerfile?

I have only used it to build the plugin. You can upload the *.zip under external modules in dolibarr after the build process.

**Comment by carstengrimm** [comment]
> > > Hello,
> > > my solution is via dockerfile.
> > > 
> > > 1. Clone repository
> > > 2. Touch dockerfile in new dir
> > > 3. Make subdir "app"
> > > 4. Copy reposiroty into the workdir "app"
> > > 5. Build docker file
> > > 6. Copy the plugin *.zip via docker exec from /tmp to the workdir /app
> > 
> > 
> > can you elaborate a bit more?
> > ```
> > * you provide a dolibarr installation into the package??
> > 
> > * or are you adding the "app" folder and dockerfile to your dolibarr root then run dockerfile?
> > ```
> 
> I have only used it to build the plugin. You can upload the *.zip under external modules in dolibarr after the build process.


I'm trying to build the zip file, but whether or not I use windows or linux, I am running into error the following error. my setup is like this:

- folder/dockerfile
- folder/app
- folder/app/pluginfiles.ext
- folder/app/conposer.json

```
------
 > [7/8] RUN composer install --no-interaction --prefer-dist:
0.442 Composer plugins have been disabled for safety in this non-interactive session. Set COMPOSER_ALLOW_SUPERUSER=1 if you want to allow plugins to run as root/super user.
0.442 Do not run Composer as root/super user! See https://getcomposer.org/root for details
0.447 Composer could not find a composer.json file in /app
0.447 To initialize a project, please create a composer.json file. See https://getcomposer.org/basic-usage
------
dockerfile:36
--------------------
  34 |
  35 |     # Install dependencies with composer
  36 | >>> RUN composer install --no-interaction --prefer-dist
  37 |
  38 |     # Run make zip
--------------------
ERROR: failed to solve: process "/bin/sh -c composer install --no-interaction --prefer-dist" did not complete successfully: exit code: 1
exit status 1

```

seems it's searching in the wrong place for the composer file

**Comment by ultrasites** [comment]
> > > > Hello,
> > > > my solution is via dockerfile.
> > > > 
> > > > 1. Clone repository
> > > > 2. Touch dockerfile in new dir
> > > > 3. Make subdir "app"
> > > > 4. Copy reposiroty into the workdir "app"
> > > > 5. Build docker file
> > > > 6. Copy the plugin *.zip via docker exec from /tmp to the workdir /app
> > > 
> > > 
> > > can you elaborate a bit more?
> > > ```
> > > * you provide a dolibarr installation into the package??
> > > 
> > > * or are you adding the "app" folder and dockerfile to your dolibarr root then run dockerfile?
> > > ```
> > 
> > 
> > I have only used it to build the plugin. You can upload the *.zip under external modules in dolibarr after the build process.
> 
> I'm trying to build the zip file, but whether or not I use windows or linux, I am running into error the following error. my setup is like this:
> 
>     * folder/dockerfile
> 
>     * folder/app
> 
>     * folder/app/pluginfiles.ext
> 
>     * folder/app/conposer.json
> 
> 
> ```
> ------
>  > [7/8] RUN composer install --no-interaction --prefer-dist:
> 0.442 Composer plugins have been disabled for safety in this non-interactive session. Set COMPOSER_ALLOW_SUPERUSER=1 if you want to allow plugins to run as root/super user.
> 0.442 Do not run Composer as root/super user! See https://getcomposer.org/root for details
> 0.447 Composer could not find a composer.json file in /app
> 0.447 To initialize a project, please create a composer.json file. See https://getcomposer.org/basic-usage
> ------
> dockerfile:36
> --------------------
>   34 |
>   35 |     # Install dependencies with composer
>   36 | >>> RUN composer install --no-interaction --prefer-dist
>   37 |
>   38 |     # Run make zip
> --------------------
> ERROR: failed to solve: process "/bin/sh -c composer install --no-interaction --prefer-dist" did not complete successfully: exit code: 1
> exit status 1
> ```
> 
> seems it's searching in the wrong place for the composer file

Yes, the dockerfile is wrong 

Here is an update. Build it `docker build -t YOUR_IMAGE_NAME .` and run it via docker with a volume `docker run -it -v $(pwd)/app:/app YOUR_IMAGE_NAME`.  Execute `composer i` and `make zip` in the container.

```
FROM debian:12.9

# Install necessary dependencies
RUN apt-get update && apt-get install -y \
    curl \
    unzip \
    php-cli \
    php-gd \
    php-xml \
    php-zip \
    composer \
    make \
    && rm -rf /var/lib/apt/lists/*

# Install php-scoper
RUN curl -L -o /usr/local/bin/php-scoper \
    https://github.com/humbug/php-scoper/releases/download/0.18.11/php-scoper.phar \
    && chmod +x /usr/local/bin/php-scoper

# Verify installation
RUN php-scoper --version

# Define a volume for persistent storage
VOLUME ["/app"]

# Set working directory
WORKDIR /app

```

**Troubleshooting**

I have removed Line 24 in the makefile. `(cd "/tmp/${MODULE}-scoper" && ${PHPSTANCMD} analyse --error-format gitlab --no-progress -c .phpstan.local.neon --memory-limit 6G)`. PHPStan is an analyzer. The repo config for phpstan has pathes to a wrong place. Code analysis is not needed in this case, in my opinion.

**Comment by carstengrimm** [comment]
> FROM debian:12.9
> 
> # Install necessary dependencies
> RUN apt-get update && apt-get install -y \
>     curl \
>     unzip \
>     php-cli \
>     php-gd \
>     php-xml \
>     php-zip \
>     composer \
>     make \
>     && rm -rf /var/lib/apt/lists/*
> 
> # Install php-scoper
> RUN curl -L -o /usr/local/bin/php-scoper \
>     https://github.com/humbug/php-scoper/releases/download/0.18.11/php-scoper.phar \
>     && chmod +x /usr/local/bin/php-scoper
> 
> # Verify installation
> RUN php-scoper --version
> 
> # Define a volume for persistent storage
> VOLUME ["/app"]
> 
> # Set working directory
> WORKDIR /app

I was able to create the plugin zip, however there are still a few things:

```
(cd "/tmp/facturx-scoper/build" && ./cleanup_vendor.sh)
bash: line 1: ./cleanup_vendor.sh: Permission denied
make: *** [Makefile.dist:27: scoperzip] Error 126
root@6b84574c13a0:/app# exit
```
for testing purposes since I could not change the permissions to make it work, I have removed the ./cleanup_vendor.sh command and the filesize is getting over 5 MB which makes the zip uploader of dolibarr not work - but at least the zip build completes. Instead of the zip uploader, I then just uploaded the plugin to custom/facturx/ and it appeared in the list of available modules to 

also I had to change a few things because the path's were not correct... 

In Makefile:
  PHPCMD = php8.2 
  COMPOSERCMD = /usr/bin/composer

In Makefile.dist:

	/usr/local/bin/php-scoper add-prefix --output-dir "/tmp/${MODULE}-scoper" --config .scoper.inc.php --force --ansi


and the output still shows:

```
module name = facturx, version = 1.6.64
module archive is ready : /tmp/module_facturx-1.6.64.zip ...
make: warning:  Clock skew detected.  Your build may be incomplete.
```

**Comment by FrankieGoestoHollywood** [comment]
Hi, I’m new to Peppol, so please excuse me if I’ve misunderstood anything.

I’m from Belgium, where Peppol will become mandatory for B2B transactions starting January 1st, 2026. It is already mandatory for government invoicing.

Thanks @rycks for your efforts and for your statement about making it free "when it becomes mandatory" (Is this for government invoicing? B2B? B2C?). Currently, the module is being sold for €500. For me, this seems to defeat the purpose of using Dolibarr, but I understand you need to make a living.

The module description mentions it has been “confirmed by many Belgians who use it to import Dolibarr invoices into WinBooks.” This feels a bit vague - How many and is it just for WinBooks?

From what I understand, Peppol is the format, and it generates an XML file. So far, I haven’t found a module that zips this XML file for testing purposes. I’m not a developer, so I haven’t been able to build one myself. If someone could publish a solution without charging a fortune, that would be great.

However, you also need to send the XML to an Access Point, like WinBooks. Typically, this requires a paid service. Do you know of any Access Points that will be supported? Are any free, like Billit?

If the approach is still being debated, I believe one module should ideally support multiple Access Points.

My accountant is proposing a €15/month fee to use his “soon-to-be-mandatory” accounting system. 
I’d prefer to be ready by the end of the year with a solution that costs the same as I’ve paid so far: nothing.

Given that Peppol is being adopted worldwide and becoming increasingly mandatory, would it be possible to integrate it directly into Dolibarr, similar to how other modules can be 'enabled'?

I’d appreciate hearing when something usable might be available, or I’ll have to start considering alternatives.
Sorry for the long post and again, many thanks for your efforts so far.

**Comment by rycks** [comment]
@FrankieGoestoHollywood be carefull peppol is not the same thing as facturx  ... and that github issue is not the right place to make business exchanges, please contribute with code :-)

(just a parenthesis ... for you 500€ is too much, nice, i was a test but when the module was at 150€ you don't buy it too and now as i'm working on autolink with peppol ap. So i spend a lot of time and nobody pay for that so the problem loops to myself : "maybe customers" says "i will wait" and "you have to do it (for no money)" ... please consider that https://cap-rel.fr/services/soutien-rd/ and send me a private mail i'm open to make gifts as all of my contacts knows)

**Comment by FrankieGoestoHollywood** [comment]
I agree that GitHub shouldn’t be used to promote your business. However, “requests for new features are managed on GitHub.”

That said, I think the whole approach to handling the Peppol obligation feels incomplete.

This directive has led to literally hundreds of Access Point providers, all competing to make money.

It reminds me of the situation with SSL certificates before ‘Let’s Encrypt’ appeared.

What’s really needed is a free alternative to these providers.

Paying both for a module and for an Access Point provider only perpetuates this outdated ecosystem.

Is anyone aware of any initiatives aiming to replace these Access Points, similar to what Let’s Encrypt did for SSL?

**Comment by rycks** [comment]
@FrankieGoestoHollywood everything is here https://www.dolibarr.fr/forum/t/giff-facture-electronique/46880

in brief, some of us are working together to find a global solution

**Comment by marclaporte** [comment]
@sonikf:

> Countries outside Europe such as Australia, Canada, China, India, Japan, USA(that's about half the world population)...all using Peppol.
> You can see the full list [here](https://peppol.org/members/peppol-certified-service-providers/)

That is a list of service providers and the page indicates: "The information regarding the SP Country / Territory reflects the country / territory of legal residence, not the area(s) of operations."

I am in Canada and I can assure you Peppol is not used here. You can search in the official site of the Canadian government and see that the term Peppol yields "No pages were found that match your search terms."
https://www.canada.ca/

As for China, India and the USA (which together count over 3 billion people): They are not listed on peppol.org and Wikipedia. Thus, I conclude your statement is incorrect.
- 20 countries are listed here: https://peppol.org/learn-more/country-profiles/
- 16 countries are listed here: https://en.wikipedia.org/wiki/PEPPOL

I am not taking a position on if Peppol support should be or not be built into Dolibarr. I am just pointing out incorrect infomation.

**Comment by sonikf** [comment]
Hi @marclaporte
Thanks for pointing that out.
Although my statement was used as a hyperbole to point the adoption rate of both formats discussed regarding cross border trade, you are right that is incorrect and i have amended the original post for clarity.

**Comment by facturx-engine** [comment]
Hi everyone,

I've been following this thread for a while and wanted to share an architectural perspective that might help unblock the current situation.

The core difficulty here isn't just about generating the XML — it's about the **binary-level assembly** required by EN 16931 / ZUGFeRD. Specifically, the standard requires:

1. Embedding the CII or UBL XML as an **attachment inside the PDF binary** (not as a separate file in an email), with a precise `AFRelationship = Alternative` flag in the PDF dictionary.
2. Writing the **XMP metadata block** (`fx:ConformanceLevel`, `fx:DocumentType`, the `pdfaExtension` schema) into the PDF's XMP stream.
3. Ensuring the output validates as **PDF/A-3b** — a constraint that most PHP PDF libraries were never designed for.

This is why the "separate PDF + XML attachment" workaround keeps appearing in Dolibarr threads: it's the path of least resistance, but it breaks the reception workflow in tools like Sage, Datev, or Lexware, which expect a **single hybrid file** and parse the XML attachment priority-first (as mandated by the German authorities).

---

### A practical solution: Docker sidecar for ZUGFeRD generation

Rather than trying to solve this inside Dolibarr's PHP stack (which would require a massive, fragile native implementation), one approach that integrates cleanly is to delegate the PDF/A-3 assembly to a **local Docker microservice** that Dolibarr calls via a simple HTTP request.

The flow becomes:

```
Dolibarr (PHP) → generates invoice data (JSON)
     ↓
POST http://localhost:8000/v1/xml          → returns EN 16931 CII XML
POST http://localhost:8000/v1/merge        → takes your existing PDF + the XML → returns ZUGFeRD PDF/A-3b
```

Example cURL (or equivalent in PHP with Guzzle):

```bash
# Step 1: Generate the EN 16931 XML from your ERP data
curl -X POST "http://localhost:8000/v1/xml" \
  -H "Content-Type: application/json" \
  -d '{
    "invoice_number": "INV-2025-001",
    "issue_date": "20250901",
    "seller": { "name": "Your Company", "vat_number": "FR12345678901" },
    "buyer": { "name": "Client GmbH" },
    "lines": [{ "name": "Services", "net_price": 1000.00, "vat_rate": 20.0 }],
    "amounts": { "grand_total": "1200.00" }
  }' -o invoice.xml

# Step 2: Merge into your existing PDF (your current Dolibarr-generated PDF)
curl -X POST "http://localhost:8000/v1/merge" \
  -F "pdf=@/path/to/dolibarr_invoice.pdf" \
  -F "xml=@invoice.xml" \
  --output facturx_compliant.pdf
```

The local engine runs **100% air-gapped** — no data leaves your server. Validation is performed by the same Schematron engine (SaxonC-HE) used by KoSIT (Germany) and Chorus Pro (France), so you get a technical parity guarantee with the official portals.

An **interactive demo** is available here if you want to test it with your own files before setting up anything: 🤗 <https://huggingface.co/spaces/Facturx-engine/factur-x-engine-demo>

And the Docker image (community edition, free / MIT): <https://hub.docker.com/r/facturxengine/facturx-engine>

```bash
docker run -d -p 8000:8000 facturxengine/facturx-engine:latest
```

Happy to answer technical questions about the integration pattern. The `/v1/merge` endpoint was specifically designed for the "bring your own PDF" use case, which fits perfectly with Dolibarr's existing PDF generation flow.

**Comment by NoelzeN** [comment]
> Happy to answer technical questions about the integration pattern. The `/v1/merge` endpoint was specifically designed for the "bring your own PDF" use case, which fits perfectly with Dolibarr's existing PDF generation flow.

Is my understanding correct that one would have to build a Dolibarr Plugin in order to call these API Endpoints from the process of Invoice Generation within Dolibarr and no such plugin exists (yet)? I might then look into this,

**Comment by solarpush** [comment]
FYI, we built a Facture-X generator in Go + Typst that could simplify the Dolibarr integration side. Instead of fighting PHP to produce compliant PDF/A-3 (as @rycks  said, "it's a hard job"), the PDF generation work can be offloaded to an external microservice.

Benefit: no need to maintain the PDF generation chain inside Dolibarr anymore, the service handles EN16931 + EXTENDED natively, at ~165 PDF/s.

Open source: https://github.com/solarpush/fX
